File: dotnet-backend/KapProjeBackend/Models/NewsItem.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace KapProjeBackend.Models;

/// <summary>
/// MongoDB'deki haber dökümanını temsil eder
/// </summary>
[BsonIgnoreExtraElements]
public class NewsItem
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string? Id { get; set; }

    [BsonElement("primary_ticker")]
    public string? PrimaryTicker { get; set; }

    [BsonElement("publisher_ticker")]
    public string? PublisherTicker { get; set; }

    [BsonElement("related_tickers")]
    public List<string>? RelatedTickers { get; set; }

    [BsonElement("published_at")]
    public PublishedAt? PublishedAt { get; set; }

    [BsonElement("category")]
    public string? Category { get; set; }

    [BsonElement("newsworthiness")]
    public double Newsworthiness { get; set; }

    [BsonElement("headline")]
    public string? Headline { get; set; }

    [BsonElement("facts")]
    public List<Fact>? Facts { get; set; }

    [BsonElement("key_numbers")]
    public KeyNumbers? KeyNumbers { get; set; }

    [BsonElement("tweet")]
    public TweetData? Tweet { get; set; }

    [BsonElement("seo")]
    public SeoData? Seo { get; set; }

    [BsonElement("visual_prompt")]
    public string? VisualPrompt { get; set; }

    [BsonElement("publish_target")]
    public string? PublishTarget { get; set; }

    [BsonElement("url")]
    public string? Url { get; set; }

    [BsonElement("ticker")]
    public string? Ticker { get; set; }

    [BsonElement("_inserted_at")]
    public string? InsertedAt { get; set; }
}

[BsonIgnoreExtraElements]
public class PublishedAt
{
    [BsonElement("date")]
    public string? Date { get; set; }

    [BsonElement("time")]
    public string? Time { get; set; }

    [BsonElement("timezone")]
    public string? Timezone { get; set; }
}

[BsonIgnoreExtraElements]
public class Fact
{
    [BsonElement("k")]
    public string? Key { get; set; }

    [BsonElement("v")]
    public string? Value { get; set; }
}

[BsonIgnoreExtraElements]
public class KeyNumbers
{
    [BsonElement("amount_raw")]
    public string? AmountRaw { get; set; }

    [BsonElement("ratio_to_market_cap")]
    public string? RatioToMarketCap { get; set; }

    [BsonElement("ratio_to_revenue")]
    public string? RatioToRevenue { get; set; }
}

[BsonIgnoreExtraElements]
public class TweetData
{
    [BsonElement("text")]
    public string? Text { get; set; }

    [BsonElement("hashtags")]
    public List<string>? Hashtags { get; set; }

    [BsonElement("disclaimer")]
    public string? Disclaimer { get; set; }
}

[BsonIgnoreExtraElements]
public class SeoData
{
    [BsonElement("title")]
    public string? Title { get; set; }

    [BsonElement("meta_description")]
    public string? MetaDescription { get; set; }

    [BsonElement("article_md")]
    public string? ArticleMd { get; set; }
}

File: kap-mobile/lib/models/news_item.dart
class NewsItem {
  final String? id;
  final String? primaryTicker;
  final String? publisherTicker;
  final List<String>? relatedTickers;
  final PublishedAt? publishedAt;
  final String? category;
  final double newsworthiness;
  final String? headline;
  final List<Fact>? facts;
  final String? visualPrompt;
  final String? url;
  final String? ticker;
  final Seo? seo;
  final Tweet? tweet;

  NewsItem({
    this.id,
    this.primaryTicker,
    this.publisherTicker,
    this.relatedTickers,
    this.publishedAt,
    this.category,
    this.newsworthiness = 0.0,
    this.headline,
    this.facts,
    this.visualPrompt,
    this.url,
    this.ticker,
    this.seo,
    this.tweet,
  });

  factory NewsItem.fromJson(Map<String, dynamic> json) {
    return NewsItem(
      id: json['id'] ?? json['_id'],
      primaryTicker: json['primaryTicker'] ?? json['primary_ticker'],
      publisherTicker: json['publisherTicker'] ?? json['publisher_ticker'],
      relatedTickers: (json['relatedTickers'] ?? json['related_tickers']) != null
          ? List<String>.from(json['relatedTickers'] ?? json['related_tickers'])
          : null,
      publishedAt: (json['publishedAt'] ?? json['published_at']) != null
          ? PublishedAt.fromJson(json['publishedAt'] ?? json['published_at'])
          : null,
      category: json['category'],
      newsworthiness: (json['newsworthiness'] ?? 0).toDouble(),
      headline: json['headline'],
      facts: json['facts'] != null
          ? (json['facts'] as List).map((f) => Fact.fromJson(f)).toList()
          : null,
      visualPrompt: json['visualPrompt'] ?? json['visual_prompt'],
      url: json['url'],
      ticker: json['ticker'],
      seo: json['seo'] != null ? Seo.fromJson(json['seo']) : null,
      tweet: json['tweet'] != null ? Tweet.fromJson(json['tweet']) : null,
    );
  }

  String get displayTicker => primaryTicker ?? ticker ?? 'N/A';
  
  String get displayTime {
    if (publishedAt?.time != null) {
      return publishedAt!.time!.substring(0, 5);
    }
    return '';
  }
  
  String get description {
    // Kullanıcı isteği üzerine öncelikle tweet text'ini gösteriyoruz
    if (tweet?.text != null && tweet!.text!.isNotEmpty) {
      return tweet!.text!;
    }
    if (seo?.articleMd != null && seo!.articleMd!.isNotEmpty) {
      return seo!.articleMd!;
    }
    return '';
  }

  String get summary {
    if (facts != null && facts!.isNotEmpty) {
      final validFacts = facts!
          .where((f) => f.value != null && f.value!.isNotEmpty)
          .take(2)
          .map((f) => f.value!)
          .toList();
      return validFacts.join(' ');
    }
    return '';
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'primaryTicker': primaryTicker,
      'publisherTicker': publisherTicker,
      'relatedTickers': relatedTickers,
      'publishedAt': publishedAt != null ? {
        'date': publishedAt!.date,
        'time': publishedAt!.time,
        'timezone': publishedAt!.timezone,
      } : null,
      'category': category,
      'newsworthiness': newsworthiness,
      'headline': headline,
      'facts': facts?.map((f) => {'k': f.key, 'v': f.value}).toList(),
      'visualPrompt': visualPrompt,
      'url': url,
      'ticker': ticker,
      'seo': seo != null ? {
        'title': seo!.title,
        'meta_description': seo!.metaDescription,
        'article_md': seo!.articleMd,
      } : null,
      'tweet': tweet != null ? {
        'text': tweet!.text,
        'hashtags': tweet!.hashtags,
        'disclaimer': tweet!.disclaimer,
      } : null,
    };
  }
}

class PublishedAt {
  final String? date;
  final String? time;
  final String? timezone;

  PublishedAt({this.date, this.time, this.timezone});

  factory PublishedAt.fromJson(Map<String, dynamic> json) {
    return PublishedAt(
      date: json['date'],
      time: json['time'],
      timezone: json['timezone'],
    );
  }
}

class Fact {
  final String? key;
  final String? value;

  Fact({this.key, this.value});

  factory Fact.fromJson(Map<String, dynamic> json) {
    return Fact(
      key: json['key'] ?? json['k'],
      value: json['value'] ?? json['v'],
    );
  }
}

class Seo {
  final String? title;
  final String? metaDescription;
  final String? articleMd;

  Seo({this.title, this.metaDescription, this.articleMd});

  factory Seo.fromJson(Map<String, dynamic> json) {
    return Seo(
      title: json['title'],
      metaDescription: json['metaDescription'] ?? json['meta_description'],
      articleMd: json['articleMd'] ?? json['article_md'],
    );
  }
}

class Tweet {
  final String? text;
  final List<String>? hashtags;
  final String? disclaimer;

  Tweet({this.text, this.hashtags, this.disclaimer});

  factory Tweet.fromJson(Map<String, dynamic> json) {
    return Tweet(
      text: json['text'],
      hashtags: json['hashtags'] != null ? List<String>.from(json['hashtags']) : null,
      disclaimer: json['disclaimer'],
    );
  }
}
